src/tools/storage-key-recipe-resolver.ts

```diff
@@ -0,0 +1,127 @@
+      resolved.normalize();
+      if (!resolved.isResolved()) {
+        throw Error(`Recipe ${resolved.name} did not properly resolve!\n${resolved.toString({showUnresolved: true})}`);
+      }
+      yield resolved;
+    }
+  }
+
+  /**
+   * Resolves unresolved recipe or normalizes resolved recipe.
+   *
+   * @param recipe long-running or ephemeral recipe
+   * @param arc Arc is associated with input recipe
+   * @param opts contains `errors` map for reporting.
+   */
+  async resolveOrNormalize(recipe: Recipe, arc: Arc, opts?: IsValidOptions): Promise<Recipe | null> {
+    const normalized = recipe.clone();
+    normalized.normalize();
+    if (normalized.isResolved()) return normalized;
+
+    return await (new RecipeResolver(arc).resolve(recipe, opts));
+  }
+
+  /** Returns the arcId from annotations on the recipe when present. */
+  getArcId(recipe: Recipe): string | null {
+    for (const trigger of recipe.triggers) {
+      for (const [key, val] of trigger) {
+        if (key === 'arcId') {
+          return val;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Create stores with keys for all create handles.
+   *
+   * @param recipe should be long running.
+   * @param arc Arc is associated with current recipe.
+   */
+  createStoresForCreateHandles(recipe: Recipe, arc: Arc) {
+    const resolver = new CapabilitiesResolver({arcId: arc.id});
+    for (const createHandle of recipe.handles.filter(h => h.fate === 'create')) {
+      const storageKey = ramDiskStorageKeyPrefixForTest()(arc.id); // TODO(#4818) create the storage keys.
+      const store = new Store({storageKey, exists: Exists.MayExist, type: createHandle.type, id: createHandle.id});
+      arc.context.registerStore(store, createHandle.tags);
+    }
+  }
+
+  /**
+   * TODO(#4818) method to match `map` and `copy` fated handles with storage keys from `create` handles.
+   *
+   * @throws when a mapped handle is associated with too many stores (ambiguous mapping).
+   * @throws when a mapped handle isn't associated with any store (no matches found).
+   * @throws when handle is mapped to a handle from an ephemeral recipe.
+   * @param recipe long-running or ephemeral recipe
+   */
+  matchKeysToHandles(recipe: Recipe) {
+    recipe.handles
+      .filter(h => h.fate === 'map' || h.fate === 'copy')
+      .forEach(handle => {
+        const matches = this.runtime.context.findHandlesById(handle.id)
+          .filter(h => h.fate === 'create');
+
+        if (matches.length === 0) {
+          throw Error(`No matching handles found for ${handle.localName}.`);
+        } else if (matches.length > 1) {
+          throw Error(`More than one handle found for ${handle.localName}.`);
+        }
+
+        const match = matches[0];
+        if (!match.recipe.isLongRunning) {
+          throw Error(`Handle ${handle.localName} mapped to ephemeral handle ${match.localName}.`);
+        }
+
+        handle.storageKey = match.storageKey;
```

Comment:
You cannot do that for 'copy' handles, for them we would need to have a machanism to require creating a new store with duplicate data. But we don't need this for March, so just remove handling of the 'copy' fate and we will not resolve is someone tries using it.

Reply:
Maria suggested that I include copy fates. My plan is to use this method more to check that handles obey the lifetime restrictions of long-running / ephemeral arcs; I'll assign storage keys some other way. I don't think that I'll get away with normal assignment to move the storage keys anyway -- I'll sync with Maria about it.